storage/redis/peer_store.go

```diff
@@ -0,0 +1,803 @@
+	IPver := p.IP.AddressFamily.String()
+	log.Debug("storage: GraduateLeecher", log.Fields{
+		"InfoHash": ih.String(),
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", p.ID.String(), p.IP.String(), IPver, p.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	encodedInfoHash := ih.String()
+	encodedLeecherInfoHash := IPver + "_L_" + encodedInfoHash
+	encodedSeederInfoHash := IPver + "_S_" + encodedInfoHash
+
+	_, err := conn.Do("HDEL", encodedLeecherInfoHash, pk)
+	if err != nil {
+		return err
+	}
+
+	// Update the peer in the swarm.
+	ct := ps.getClock()
+	_, err = conn.Do("HSET", encodedSeederInfoHash, pk, ct)
+	if err != nil {
+		return err
+	}
+	_, err = conn.Do("HSET", IPver, encodedSeederInfoHash, ct)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ps *peerStore) AnnouncePeers(ih bittorrent.InfoHash, seeder bool, numWant int, announcer bittorrent.Peer) (peers []bittorrent.Peer, err error) {
+	IPver := announcer.IP.AddressFamily.String()
+	log.Debug("storage: AnnouncePeers", log.Fields{
+		"InfoHash": ih.String(),
+		"seeder":   seeder,
+		"numWant":  numWant,
+		"Peer":     fmt.Sprintf("[ID: %s, IP: %s(AddressFamily: %s), Port %d]", announcer.ID.String(), announcer.IP.String(), IPver, announcer.Port),
+	})
+
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	encodedInfoHash := ih.String()
+	encodedLeecherInfoHash := IPver + "_L_" + encodedInfoHash // key
+	encodedSeederInfoHash := IPver + "_S_" + encodedInfoHash  // key
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechers, err := conn.Do("HKEYS", encodedLeecherInfoHash)
+	if err != nil {
+		return nil, err
+	}
+	conLeechers := leechers.([]interface{})
+
+	seeders, err := conn.Do("HKEYS", encodedSeederInfoHash)
+	if err != nil {
+		return nil, err
+	}
+	conSeeders := seeders.([]interface{})
+
+	if len(conLeechers) == 0 && len(conSeeders) == 0 {
+		return nil, storage.ErrResourceDoesNotExist
+	}
+
+	if seeder {
+		// Append leechers as possible.
+		for _, pk := range conLeechers {
+			if numWant == 0 {
+				break
+			}
+
+			peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+			numWant--
+		}
+	} else {
+		// Append as many seeders as possible.
+		for _, pk := range conSeeders {
+			if numWant == 0 {
+				break
+			}
+
+			peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+			numWant--
+		}
+
+		// Append leechers until we reach numWant.
+		if numWant > 0 {
+			announcerPK := newPeerKey(announcer)
+			for _, pk := range conLeechers {
+				if pk == announcerPK {
+					continue
+				}
+
+				if numWant == 0 {
+					break
+				}
+
+				peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+				numWant--
+			}
+		}
+	}
+
+	APResult := ""
```

Comment:
What is this entire block for? If you want a string representation of the last peer in the list, use

```go
if len(peers) > 0 {
    lastPeerStr = peers[len(peers)-1].String()
    // use lastPeerStr...
}
```

Reply:
That will build a slice of strings on every announce, regardless of whether debug logging is turned on.

You actually don't need to log this at all here. The logic logs the generated announce response if debug logging is turned on, see https://github.com/chihaya/chihaya/blob/master/middleware/logic.go#L62

Reply:
OK, I will remove it.